mantidproject/mantid
# Enable Ninja Builds in Visual Studio

**Description of work.**

Allow a ninja build including use of ccache to be run from within Visual Studio 2019 or 2022 using the CMake integration features. The main change to allow this was increasing the version number in the CMakePresets.json file from 1 to 2 which then means Visual Studio parses it. I've made some other related changes:

- clarified the use of MSVC in the dev docs. Both the ninja and Visual Studio generators use the MSVC compiler on Windows so I've changed the name of the win-msvc preset to win-vs
- changed the dev docs to suggest new starters install Visual Studio 2022 instead of 2019. Both are still supported but seems sensible now to switch to the newer version
- increased the minimum required cmake version in the CMakePresets.json file to 3.21 (which is the version where the VS 2022 generator is available)
- Removed the logic that only allocated targets to folders for the Visual Studio generator. Folders specified when using the Ninja generator are read when the resulting project is opened in Visual Studio
- fixed a problem where running with CONDA_ENV=false followed by CONDA_ENV=true in the same source directory caused an error during the cmake configure step. I've set the Python_EXECUTABLE cmake variable explicitly in the CONDA_ENV=true logic. This switch seems more likely to happen when using the CMake targets view in Visual Studio where the default preset selected can be something other than "win-ninja"

The general experience in Visual Studio is approximately similar to what you get when using the Visual Studio cmake generator. The main difference is that you can't start up workbench by setting the "workbench" target as a start up item. So I've been running workbench from PyCharm instead. I think whatever configuration allows this to work with the Visual Studio generator isn't available with the ninja set-up. Also you can't right click on a folder (eg MantidFramework) and build all the targets inside the folder. I've seen online that this is available in a preview version of Visual Studio but it's not in my version of VS2022 (which is v17.4.2)

**To test:**

Probably worth cloning mantid into a new source code directory in case this doesn't work for any reason and trashes your main directory: `git clone https://github.com/mantidproject/mantid.git mantidninja`

Follow the dev-docs instructions in:

a) the CCache parts of BuildingWithCMake.html
b) and the updated GettingStartedCondaWindows.html page - run the ninja cmake option using `cmake --preset=win-ninja -B build` and then choose the "Compile and Build using Ninja" option

Open up the Ninja configured build in Visual Studio with the solution explorer set to "CMake Targets View" and the Configuration set to "win-ninja". The target system\configuration\build dropdowns should look like:

![image](https://user-images.githubusercontent.com/56295817/207627982-381d9a79-7221-4d4e-9f70-e0e608ba0aef.png)

...and the solution explorer should look like:

![image](https://user-images.githubusercontent.com/56295817/207627110-4b716c77-89e6-44ff-be36-1351d6ff8880.png)

1) Select Build, Build All. This will build everything except the tests
2) Once this completes check the incremental build is working OK by repeating and observing that no work is done
3) Right click on the "Mantid Project" entry in the tree and select "Delete cache and Reconfigure" 
4) Select Build, Build All. This should use CCache to rebuild the code more quickly than 1)

*There is no associated issue.*

*This does not require release notes* because **there's no functional change to end users, only affects developers**



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
